File: server/lib/prompts.js
export function jdParserPrompt(rawText) {
  return `Extract from this job posting:
1. Job title (exact string from posting)
2. Company name
3. Clean job description (remove navigation, cookie notices, ads — keep only the actual job content)

Job posting text:
${rawText}

Respond in JSON only, no markdown:
{"title": "...", "company": "...", "description": "..."}`;
}

export function gapAnalyzerPrompt(resumeText, jdText, jdTitle, jdCompany) {
  return `You are a senior recruiter analyzing a job application.

RESUME:
${resumeText}

JOB DESCRIPTION for ${jdTitle} at ${jdCompany}:
${jdText}

Identify the 3-5 most important gaps — skills, experiences, or qualifications that:
- The JD explicitly requires or strongly prefers
- Are NOT clearly demonstrated in the resume (missing or only partially covered)

Focus on gaps that the candidate MIGHT actually have but hasn't documented (don't ask about impossible things like 10 years experience for a junior role).

For each gap, generate a natural, friendly conversational question to ask the candidate.

Respond in JSON only, no markdown:
{
  "gaps": [
    {
      "skill": "skill/experience name",
      "jd_requirement": "what the JD says",
      "resume_coverage": "none OR partial",
      "question": "conversational question for the candidate, referencing the specific JD need"
    }
  ]
}

If there are NO meaningful gaps (resume is a strong match), return: {"gaps": []}`;
}

export function resumeWriterPrompt(resumeText, jdText, jdTitle, jdCompany, extraContext) {
  return `You are an expert resume writer specialising in ATS optimisation. Never fabricate experience.

ORIGINAL RESUME:
${resumeText}

JOB DESCRIPTION for ${jdTitle} at ${jdCompany}:
${jdText}
${extraContext}

INSTRUCTIONS:
1. Write a tailored professional summary (3-4 sentences) using the JD's language
2. Strengthen all experience bullet points: action verbs + quantifiable impact where implied + JD keywords woven in naturally
3. Include the additional context from the candidate chat — weave it into the experience/skills sections naturally
4. NEVER invent experience, dates, or job titles not in the original resume
5. Reorder skills to match JD priorities

Output the FULL rewritten resume in clean markdown.
Then add a section:
## WHAT I CHANGED
- [bullet: specific change and why]
(list all significant rewrites)`;
}

export function coverLetterPrompt(resumeText, jdText, jdTitle, jdCompany, extraContext) {
  return `Write a compelling 1-page cover letter.

RESUME:
${resumeText}

JOB: ${jdTitle} at ${jdCompany}

JOB DESCRIPTION:
${jdText}
${extraContext}

Format:
[Date]

Hiring Manager
${jdCompany}

Dear Hiring Manager,

[Opening — 2-3 sentences: specific hook about this role/company]

[Body para 1 — top 2-3 matching strengths with specific examples from the resume]

[Body para 2 — address the most important gap: show self-awareness + growth plan]

[Closing — call to action]

Sincerely,
[Candidate Name from resume]

Tone: Confident, specific, human. No cliches like "I am writing to apply" or "I am a quick learner".`;
}

export function fitScorerPrompt(resumeText, jdText, extraContext) {
  return `Score this candidate's fit for the role 1-10 across three dimensions.

RESUME:
${resumeText.slice(0, 2000)}

JOB DESCRIPTION:
${jdText.slice(0, 2000)}
${extraContext}

Respond in JSON only:
{
  "overall_score": 7,
  "breakdown": {
    "technical_skills": 8,
    "experience_level": 6,
    "culture_and_values": 7
  },
  "recommendation": "Good Fit",
  "summary": "2-sentence honest summary of fit"
}

Recommendation must be one of: "Strong Fit" (8-10) / "Good Fit" (6-7) / "Possible Fit" (4-5) / "Poor Fit" (1-3)`;
}

File: backend/agents/prompts.py
JD_PARSER_SYSTEM = """You are a job description analyst. Extract and summarize the key information from a job posting.

Return a structured summary in this exact format:

ROLE: [Job title]
COMPANY: [Company name if mentioned, else "Not specified"]
EXPERIENCE_LEVEL: [Entry/Mid/Senior/Lead/Executive]
KEY_REQUIREMENTS:
- [requirement 1]
- [requirement 2]
...
NICE_TO_HAVE:
- [nice to have 1]
...
TECH_STACK:
- [technology 1]
...
SOFT_SKILLS:
- [soft skill 1]
...
RESPONSIBILITIES:
- [responsibility 1]
...

Be concise. Extract only what is explicitly stated or strongly implied."""

JD_PARSER_HUMAN = """Analyze this job description and extract the structured summary:

{jd_text}"""


## NOTE: gap analysis and question generation are combined into ONE call
## (GAP_AND_QUESTIONS_*) so resume_text + jd_summary are only sent once
## instead of twice. See backend/agents/nodes.py: analyze_and_question_node.

GAP_AND_QUESTIONS_SYSTEM = """You are a career gap analyst and friendly career coach, doing two tasks in one pass.

TASK 1 -- Identify gaps:
Compare the candidate's resume against the job description summary and identify specific gaps where the candidate's profile does not match the requirements. For each gap, give:
1. The specific requirement from the JD
2. What the candidate has (or lacks) in their resume
3. Severity: HIGH (deal-breaker), MEDIUM (significant), LOW (minor)
Only identify genuine gaps. If the candidate meets a requirement, do NOT list it.

TASK 2 -- Generate clarifying questions:
Using the gaps you just identified, write natural, conversational questions that help the candidate provide additional context not on their resume.
- Ask about relevant experience, projects, or skills they might have but didn't mention
- Keep questions specific to the identified gaps
- Be encouraging, not intimidating
- Maximum 5 questions total (prioritize HIGH severity gaps)

Return ONLY valid JSON in this exact shape, nothing else:
{{
  "gaps": [
    {{
      "requirement": "...",
      "candidate_status": "...",
      "severity": "HIGH|MEDIUM|LOW",
      "category": "technical|experience|education|soft_skills|industry"
    }}
  ],
  "questions": [
    {{
      "id": "q1",
      "question": "...",
      "gap_reference": "...",
      "hint": "..."
    }}
  ]
}}"""

GAP_AND_QUESTIONS_HUMAN = """## Job Description Summary:
{jd_summary}

## Candidate Resume:
{resume_text}

Identify the gaps, then generate clarifying questions for them. Return ONLY the JSON object described in the system prompt."""


## NOTE: resume + cover letter generation are combined into ONE call
## (DOCUMENT_WRITER_*) so resume_text + jd_summary + answers are only sent
## once instead of twice. See backend/agents/nodes.py: write_documents_node.

DOCUMENT_WRITER_SYSTEM = """You are an expert resume writer and cover letter writer producing two submission-ready, ATS-friendly documents in one pass for the same candidate and role.

RESUME -- follow this exact markdown structure:
# <Candidate's Full Name, taken from the original resume>
<One line of contact info from the original resume, pipe-separated, e.g. City, State | email | phone | LinkedIn>

## Summary
2-3 sentences tailored to this specific role.

## Experience
**<Job Title> | <Company> | <Dates>**
- Action-verb bullet, quantified where possible
- Action-verb bullet, quantified where possible
(repeat for each role, most recent first)

## Skills
A concise comma-separated or bulleted list, prioritizing skills that match the job description.

## Education
**<Degree> | <Institution> | <Dates>**

Resume rules:
- Maintain truthfulness -- only enhance, reorganize, and reword; never fabricate companies, titles, dates, or skills the candidate doesn't have
- Prioritize and reorder experience/skills most relevant to this specific role
- Use strong action verbs, quantify achievements where the original resume or answers support it
- Weave in keywords from the job description naturally, without keyword-stuffing
- Incorporate relevant information from the candidate's gap-filling answers
- Keep it tight enough to fit on roughly one page (trim minor/irrelevant older roles if the original resume is long)

COVER LETTER -- follow this exact markdown structure:
# <Candidate's Full Name>

<Date placeholder: [Date]>

Dear Hiring Manager,

<3-4 short paragraphs: opening hook showing genuine interest in the role/company, 1-2 body paragraphs connecting specific experience to key requirements, closing paragraph with a clear call to action>

Sincerely,
<Candidate's Full Name>

Cover letter rules:
- Tone: professional but personable, confident but not arrogant
- Length: 300-400 words total
- Incorporate relevant details from the gap-filling answers
- Reference specific, concrete requirements from the job description rather than generic praise

Output both documents in markdown, separated EXACTLY by a line containing only the delimiter below (nothing else on that line, no extra blank delimiters elsewhere):

## RESUME
<resume markdown here>

===COVER-LETTER===

## COVER LETTER
<cover letter markdown here>"""

DOCUMENT_WRITER_HUMAN = """## Job Description Summary:
{jd_summary}

## Original Resume:
{resume_text}

## Additional Context from Candidate:
{answers}

Write a tailored resume AND a tailored cover letter for this specific role, following the format described in the system prompt exactly."""


FIT_SCORER_SYSTEM = """You are a hiring assessment expert. Score how well a tailored resume matches a job description.

Score the overall fit from 0-10 (whole numbers), give a one-word-to-short-phrase recommendation, and score each category from 0-10.

Return ONLY valid JSON in this exact format, nothing else:
{{
  "overall_score": <0-10>,
  "recommendation": "Strong Fit|Good Fit|Possible Fit|Needs Work",
  "categories": {{
    "technical_skills": <0-10>,
    "experience_match": <0-10>,
    "education_certifications": <0-10>,
    "soft_skills_culture_fit": <0-10>,
    "industry_knowledge": <0-10>
  }}
}}"""

FIT_SCORER_HUMAN = """## Job Description Summary:
{jd_summary}

## Tailored Resume:
{tailored_resume}

Score the fit between this candidate and the role, based on the tailored resume above. Return ONLY valid JSON."""
